pandas/core/base.py

```diff
@@ -917,7 +917,13 @@ def __iter__(self):
         (for str, int, float) or a pandas scalar
         (for Timestamp/Timedelta/Interval/Period)
         """
-        return iter(self.tolist())
+        # We are explicity making element iterators.
+        if is_datetimelike(self._values):
+            return map(com._maybe_box_datetimelike, self._values)
+        elif is_extension_array_dtype(self._values):
+            return iter(self._values)
+        else:
+            return map(self._values.item, range(self._values.size))
```

Comment:
Maybe the issue is here. Because DataFrame wants Python types, we have to convert from numpy to Python types. Before, this was done inside C (inside tolist call), now it is done in Python.

So in some way, this might require numpy to have an iterator reading out Python values, implemented in C, to make this same efficiency.

Reply:
I would say that even leaving aside language differences, we are now making ``len(self._values)`` separate calls rather than a single one. We know numpy indexing has overhead, and I think this is perfectly acceptable.